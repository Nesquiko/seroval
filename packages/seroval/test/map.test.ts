/* eslint-disable no-await-in-loop */
import { describe, it, expect } from 'vitest';
import {
  AsyncServerValue,
  deserialize,
  Feature,
  fromJSON,
  serialize,
  serializeAsync,
  ServerValue,
  toJSON,
  toJSONAsync,
} from '../src';

describe('Map', () => {
  describe('serialize', () => {
    it('supports Map', () => {
      const example = new Map([[1, 2], [3, 4]]);
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Map<number, number>>(result);
      expect(back).toBeInstanceOf(Map);
      expect(back.get(1)).toBe(example.get(1));
      expect(back.get(3)).toBe(example.get(3));
    });
    it('supports self-recursion', () => {
      const example: Map<ServerValue, ServerValue> = new Map();
      example.set(example, example);
      const result = serialize(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Map<ServerValue, ServerValue>>(result);
      expect(back.has(back)).toBe(true);
      expect(back.get(back)).toBe(back);
    });
  });
  describe('serializeAsync', () => {
    it('supports Map', async () => {
      const example = new Map([[1, 2], [3, 4]]);
      const result = await serializeAsync(Promise.resolve(example));
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<Map<number, number>>>(result);
      expect(back).toBeInstanceOf(Map);
      expect(back.get(1)).toBe(example.get(1));
      expect(back.get(3)).toBe(example.get(3));
    });
    it('supports self-recursion', async () => {
      const example: Map<AsyncServerValue, AsyncServerValue> = new Map();
      example.set(Promise.resolve(example), Promise.resolve(example));
      const result = await serializeAsync(example);
      expect(result).toMatchSnapshot();
      const back = deserialize<Map<Promise<any>, Promise<any>>>(result);
      for (const [key, value] of back) {
        expect(await key).toBe(back);
        expect(await value).toBe(back);
      }
    });
  });
  describe('toJSON', () => {
    it('supports Map', () => {
      const example = new Map([[1, 2], [3, 4]]);
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Map<number, number>>(result);
      expect(back).toBeInstanceOf(Map);
      expect(back.get(1)).toBe(example.get(1));
      expect(back.get(3)).toBe(example.get(3));
    });
    it('supports self-recursion', () => {
      const example: Map<ServerValue, ServerValue> = new Map();
      example.set(example, example);
      const result = toJSON(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Map<ServerValue, ServerValue>>(result);
      expect(back.has(back)).toBe(true);
      expect(back.get(back)).toBe(back);
    });
  });
  describe('toJSONAsync', () => {
    it('supports Map', async () => {
      const example = new Map([[1, 2], [3, 4]]);
      const result = await toJSONAsync(Promise.resolve(example));
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<Promise<Map<number, number>>>(result);
      expect(back).toBeInstanceOf(Map);
      expect(back.get(1)).toBe(example.get(1));
      expect(back.get(3)).toBe(example.get(3));
    });
    it('supports self-recursion', async () => {
      const example: Map<AsyncServerValue, AsyncServerValue> = new Map();
      example.set(Promise.resolve(example), Promise.resolve(example));
      const result = await toJSONAsync(example);
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<Map<Promise<any>, Promise<any>>>(result);
      for (const [key, value] of back) {
        expect(await key).toBe(back);
        expect(await value).toBe(back);
      }
    });
  });
  describe('compat', () => {
    it('should throw an error for unsupported target', () => {
      expect(() => serialize(new Map(), {
        disabledFeatures: Feature.Map,
      })).toThrowErrorMatchingSnapshot();
    });
  });
  describe('compat#toJSON', () => {
    it('should throw an error for unsupported target', () => {
      expect(() => toJSON(new Map(), {
        disabledFeatures: Feature.Map,
      })).toThrowErrorMatchingSnapshot();
    });
  });
});
