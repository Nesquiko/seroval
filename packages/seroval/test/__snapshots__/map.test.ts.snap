// Vitest Snapshot v1

exports[`Map > compat > should throw an error for unsupported target 1`] = `"Unsupported type \\"Map\\""`;

exports[`Map > compat#toJSON > should throw an error for unsupported target 1`] = `"Unsupported type \\"Map\\""`;

exports[`Map > serialize > supports Map 1`] = `"new Map([[1,2],[3,4]])"`;

exports[`Map > serialize > supports self-recursion 1`] = `"(h=>(h=new Map,h.set(h,h),h))()"`;

exports[`Map > serializeAsync > supports Map 1`] = `"Promise.resolve(new Map([[1,2],[3,4]]))"`;

exports[`Map > serializeAsync > supports self-recursion 1`] = `"(h=>(h=new Map([[Promise.resolve().then(()=>h),Promise.resolve().then(()=>h)]])))()"`;

exports[`Map > toJSON > supports Map 1`] = `"{\\"t\\":{\\"t\\":14,\\"i\\":0,\\"d\\":{\\"k\\":[{\\"t\\":0,\\"s\\":1},{\\"t\\":0,\\"s\\":3}],\\"v\\":[{\\"t\\":0,\\"s\\":2},{\\"t\\":0,\\"s\\":4}],\\"s\\":2}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`Map > toJSON > supports self-recursion 1`] = `"{\\"t\\":{\\"t\\":14,\\"i\\":0,\\"d\\":{\\"k\\":[{\\"t\\":10,\\"i\\":0}],\\"v\\":[{\\"t\\":10,\\"i\\":0}],\\"s\\":1}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[0]}"`;

exports[`Map > toJSONAsync > supports Map 1`] = `"{\\"t\\":{\\"t\\":18,\\"i\\":0,\\"f\\":{\\"t\\":14,\\"i\\":1,\\"d\\":{\\"k\\":[{\\"t\\":0,\\"s\\":1},{\\"t\\":0,\\"s\\":3}],\\"v\\":[{\\"t\\":0,\\"s\\":2},{\\"t\\":0,\\"s\\":4}],\\"s\\":2}}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[]}"`;

exports[`Map > toJSONAsync > supports self-recursion 1`] = `"{\\"t\\":{\\"t\\":14,\\"i\\":0,\\"d\\":{\\"k\\":[{\\"t\\":18,\\"i\\":1,\\"f\\":{\\"t\\":10,\\"i\\":0}}],\\"v\\":[{\\"t\\":18,\\"i\\":2,\\"f\\":{\\"t\\":10,\\"i\\":0}}],\\"s\\":1}},\\"r\\":0,\\"i\\":false,\\"f\\":8191,\\"m\\":[0]}"`;
